sdk-cpp/include/do_exceptions.h

```diff
@@ -1,21 +1,45 @@
 #ifndef _DELIVERY_OPTIMIZATION_DO_EXCEPTIONS_H
 #define _DELIVERY_OPTIMIZATION_DO_EXCEPTIONS_H
 
-#include <exception>
 #include <stdint.h>
 #include <system_error>
 
-#if defined(DO_INTERFACE_COM)
-#include <winerror.h>   // FAILED macro
+#if (!DO_DISABLE_EXCEPTIONS)
+#include <exception>
 #endif
 
 namespace microsoft
 {
 namespace deliveryoptimization
 {
 
+//These error macros are defined behind guards so that they do not redefine the macro for for platforms/toolchains in which they already exist
+#ifndef RETURN_RES_IF_FAILED
+#define RETURN_RES_IF_FAILED(res)  {  \
+    int32_t __res = (res);            \
+    if(FAILED(__res)) return __res;   \
+}
+#endif
+
+#ifndef FACILITY_DELIVERY_OPTIMIZATION 
+#define FACILITY_DELIVERY_OPTIMIZATION   208
+#endif
+
+#define DO_ERROR_FROM_SYSTEM_ERROR(x) (int32_t)(0xC0000000 | (0xD0 << 16) | ((int32_t)(x) & 0x0000FFFF))
+#define DO_ERROR_FROM_STD_ERROR(x) ((int32_t)(x) <= 0 ? ((int32_t)(x)) : ((int32_t) (((int32_t)(x) & 0x0000FFFF) | (FACILITY_DELIVERY_OPTIMIZATION << 16) | 0x80000000)))
+
+#ifndef FAILED
+#define FAILED(res) (((int32_t)(res)) < 0)
+#endif
+
+#ifndef SUCCEEDED
+#define SUCCEEDED(res) (((int32_t)(res)) >= 0)
+#endif
+
 enum class errc : int32_t
 {
+    s_ok                        = 0,
```

Comment:
>s_ok [](http://example.com/codeflow?start=4&length=4)

errc enum should not include a success code.

Since we have a non-throwing interface now, we should invest in a msdo::error_code class modeled on std::error_code or reuse std::error_code by supplying our own error_category object (I believe cpprest did something like this that you can borrow).

Reply:
Good point, yeah I'll look into creating a custom error code class. Even boost has boost::system::error_code

Reply:
This one can come as a later PR into the feature branch (if you take my suggestion of completing this change as piecemeal PRs).